tools/xmlschema.py

```diff
@@ -0,0 +1,462 @@
+                # 2. maxOccurs = unbounded and minOccurs = 0
+                # the case maxOccurs = unbounded and minOccurs = 1 is refactored
+                # into 1 required element in case 1 + 1 optional element in case 2
+                # bounded_elements = [el for el in elements if not el.attrib["maxOccurs"] == "unbounded"]
+                # optional_unbounded = list()
+                # for unbounded_element in [el for el in elements if el.attrib["maxOccurs"] == "unbounded"]:
+                #     unbounded_element:ElementTree.Element
+                #     min_occurs = unbounded_element.attrib["minOccurs"]
+                #     if min_occurs == "1":
+                #         required_item = deepcopy(unbounded_element)
+                #         required_item.set("maxOccurs", "1")
+                #         bounded_elements.append(required_item)
+                    
+                #     # it will be ugly, but we can try to reduce clutter
+                #     unbounded_element.attrib.pop("minOccurs")
+                #     unbounded_element.attrib.pop("maxOccurs")
+                #     optional_unbounded.append(unbounded_element)
+
+                # infinity_choice = ElementTree.Element(_to_qname("xs:choice"))
+                # infinity_choice.set("maxOccurs", "unbounded")
+                # infinity_choice.extend(optional_unbounded)
+
+                # # bounded elements may show up anywhere between unbounded optional elements.
+                # # However we can't use <xs:all> here (XSD 1.0 limitation). Instead use a choice
+                # # over sequences of unbounded choices with a permutation of bounded elements
+                # # inbetween them.
+                # container = ElementTree.Element(_to_qname("xs:choice"))
+                # for sequence in permutations(bounded_elements):
+                #     seq_container = ElementTree.Element(_to_qname("xs:sequence"))
+                #     seq_container.append(infinity_choice)
+                #     for pair in zip_longest(sequence, [], fillvalue=infinity_choice):
+                #         seq_container.extend(pair)
+                
+                #     container.append(seq_container)
+
+                # el.append(container)
+                # el.extend(attributes)
+
+                # above code appears to work, but the generated
+                # XSD is multiple GB in size. Use this for now.
+                seq_container = ElementTree.Element(_to_qname("xs:choice"))
+                seq_container.set("maxOccurs", "unbounded")
+                seq_container.extend(elements)
+                el.append(seq_container)
+                el.extend(attributes)
+
+        else:
+            el.extend(attributes)
+
+        return namespaces, el
+
+
+def setup_schema(used_ns: list, use_default_ns: bool = True) -> ElementTree.Element:
+    xsd_schema = ElementTree.Element(_to_qname("xs:schema"))
+    xsd_schema.set("version", "1.1")
```

Comment:
I'm not sure but I guess `<?xml version='1.0' encoding='UTF-8'?>` is what you meant with your comment by not supporting XSD 1.1 and thus we use `1.0`?

However, the file generated from `sensor.sdf` contains
```
<?xml version='1.0' encoding='UTF-8'?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" version="1.1" ...
```
but wouldn't it make more sense that the version number is the version of the SDF file (directory)? So that following command would set version to `1.8`:
```bash
python3 tools/xmlschema.py -i sensor.sdf -o /ws/sdformat/test_output -s /ws/sdformat/sdf/1.8
```

Reply:
> Shouldn't this be "1.0"?

You are right; it should be 1.0 here.

> but wouldn't it make more sense that the version number is the version of the SDF file (directory)?

No, because its references to different things. The first line (the XML header) specifies the version of the XML in which the schema file is written. The second line (the XSD namespace) specified the version of XSD to use when interpreting this schema file.

To specify the SDFormat version to use when validating one would (traditionally) set the version in the namespace ([here is an example](https://github.com/FirefoxMetzger/scikit-bot/blob/main/skbot/ignition/sdformat/schema/1.8/link.xsd)). I recall that there was an issue in this repo where this feature was requested before, and I am in favor of doing so. This would, however, need prior consent from one of the maintainers, and - because this is a new feature instead of a bugfix - I left it out of this PR.

Reply:
The issue you mentioned might be https://github.com/ignitionrobotics/sdformat/issues/545 - something which should be fixed as all validation is done with the files on the website that might not represent the SDF version of your file.